# 7. Class 고급

## 클래스, 인스턴스, 오브젝트 

`-` 오브젝트
- 클래스 오브젝트
- 인스턴스 오브젝트 

`-` 클래스 (=클래스 오브젝트)

`-` 인스턴스 (=인스턴스 오브젝트)

## 클래스 속성 vs 인스턴스 속성

### 예제1

In [ ]:
class Testclass1: 
    x=0 
    y=0
    def my_print(self):
        self.x += 1 
        Testclass1.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [ ]:
f=Testclass1

In [ ]:
a=Testclass1()

In [ ]:
b=f()

In [ ]:
a.my_print()

In [ ]:
b.my_print()

In [ ]:
b.my_print()

In [ ]:
a.my_print()

In [ ]:
a.my_print()

`-` 신기한점: 각 인스턴스에서 instance.my_print()를 실행한 횟수를 서로 공유하는 듯 하다. 

#### 분석

`-` 코드를 시점별로 분석해보자. 

`-` 분석을 위해서 커널을 재시작한다. 

[시점1]: Testclass1를 선언하는 시점

In [5]:
class Testclass1: 
    x=0 
    y=0
    def my_print(self):
        self.x += 1 
        Testclass1.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [7]:
dir(Testclass1)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print',
 'x',
 'y']

In [8]:
dir(a)

NameError: name 'a' is not defined

In [9]:
dir(b)

NameError: name 'b' is not defined

`–` 이 시점에는 Testclass1만이 존재한다. Testclass1를 바로 클래스 오브젝트라고 부름. 

In [11]:
Testclass1.x

0

In [12]:
Testclass1.y

0

`–` 현재시점에서는 클래스 오브젝트의 수 1개, 인스턴스 오브젝트의 수 0개, 따라서 총 오브젝트 수는 1개임. 

[시점2] 클래스에 별칭을 지정하는 시점 

In [13]:
f=Testclass1

In [14]:
f.x

0

In [15]:
f.y

0

In [16]:
Testclass1.x

0

In [17]:
Testclass1.y

0

`–` 이 시점에서 클래스 오브젝트는 2개가 있는 것 처럼 보인다. 

`-` 그렇다면 이 2개의 클래스 오브젝트는 컴퓨터의 어딘가에 저장이 되어 있을 것이다. 

`-` 구체적으로는 메모리에 저장되어있을것. 

`-` 2개의 클래스오브젝트는 서로 다른 메모리 공간에 저장되어 있을것이다.

`-` 진짜인가? 확인해보자. `id()`는 오브젝트(클래스 오브젝트, 인스턴스 오브젝트)가 저장된 메모리 주소를 확인하는 명령어이다. 

In [18]:
id(f)

93967322676384

`–` f라는 오브젝트는 93967322676384 메모리에 저장되어 있다. 

In [19]:
id(Testclass1)

93967322676384

`-` 어? 그런데 Testclass1의 오브젝트 역시 93967322676384 메모리에 저장되어 있다. 

`-` 추론: 사실 93967322676384라는 메모리공간에 저장된 어떠한 것은 동일한데, 그것을 어떤사람은 `Testclass1` 이라고 부르고 어떤사람은 `f`라고 부른다. 

`-` 이는 마치 별명이랑 비슷하다. 나라는 오브젝트를 어떤사람은 `최규빈`이라고 부르고, 어떤사람은 `팬더`라고 부른다. 부르는 이름이 2개라고 해서 나라는 오브젝트가 2개가 있는것은 아니다. 

`-` 결국 이 시점에서 클래스 오브젝트의 수는 여전히 1개라고 볼 수 있다. (인스턴스 오브젝트의 수는 0개)

[시점3]: 클래스 오브젝트로부터 인스턴스 오브젝트를 만드는 시점

In [20]:
a=Testclass1()
b=f()

In [24]:
id(Testclass1),id(f),id(a),id(b)

(93967322676384, 93967322676384, 139694857660688, 139694848860656)

`–` 이 순간에는 클래스 오브젝트 1개, 인스턴스 오브젝트 2개 존재한다. 즉 총 3개의 오브젝트가 존재한다. 

`-` 메모리주소 93967322676384 에 존재하는 오브젝트는 클래스 오브젝트이며 Testclass1 또는 f 라고 불린다. 

`-` 메모리주소 139694857660688 에 존재하는 오브젝트는 인스턴스 오브젝트이며 a라고 불린다. 

`-` 메모리주소 139694848860656 에 존재하는 오브젝트는 인스턴스 오브젝트이며 b라고 불린다. 

In [26]:
Testclass1.x, Testclass1.y

(0, 0)

In [29]:
f.x,f.y

(0, 0)

In [27]:
a.x,a.y

(0, 0)

In [28]:
b.x,b.y

(0, 0)

[시점4]

In [30]:
a.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력


In [31]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 1), (1, 1), (0, 1))

`-` 특징 
 - a.my_print()를 실행하면 a.x 의 값이 1이 증가한다. 
 - a.my_print()를 실행하면 f.y, a.y, b.y 의 값이 동시에 1이 증가한다. (공유가 되는 느낌)

[시점5]

In [34]:
b.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 2 회 출력


In [35]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 2), (1, 2), (1, 2))

[시점6]

In [36]:
b.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 3 회 출력


In [37]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 3), (1, 3), (2, 3))

[시점7]

In [38]:
a.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 4 회 출력


In [39]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 4), (2, 4), (2, 4))

[시점8]

In [40]:
a.my_print()

현재 인스턴스에서 3 회 출력
전체 인스턴스에서 총 5 회 출력


In [41]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 5), (3, 5), (2, 5))

### 예제2

`-` 아래처럼 코드를 바꿔도 잘 동작할것 같다. 

In [46]:
class Testclass2: 
    def __init__(self):
        self.x=0
        self.y=0
    def my_print(self):
        self.x += 1 
        Testclass2.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [47]:
c=Testclass2()

In [51]:
c.my_print()

AttributeError: type object 'Testclass2' has no attribute 'y'

`–` 왜 에러가 나는가? 

In [52]:
dir(Testclass2)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print']

In [53]:
dir(Testclass1)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print',
 'x',
 'y']

`-` 관찰1: Testclass2에서는 Testclass1과는 다르게 x,y가 없다. 

In [54]:
dir(c)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print',
 'x',
 'y']

`–` 관찰2: 그런데 `c`라는 인스턴스 오브젝트에서는 x,y가 있다. 

`-` 추론: `__init__`함수는 클래스 오브젝트가 만들어지는 시점에서는 실행되지 않고, 인스텐스 오브젝트가 만들어지는 시점에 실행된다.

`-` 결국 `__init__`함수의 역할은 클래스 오브젝트에서 인스턴스 오브젝트를 만든후에 초기화를 위해서 실행하는 어떠한 일련의 명령들을 묶어놓은 것에 불과하다. 

`–` 즉 위의 코드는 굳이 따지면 아래를 실행한 것과 동일하다. 

In [55]:
class Testclass2: 
#     def __init__(self):
#         self.x=0
#         self.y=0
    def my_print(self):
        self.x += 1 
        Testclass2.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [56]:
c=Testclass2()

In [58]:
c.x=0
c.y=0

`-` 이 상황에서 
```python
c.my_print()
```
를 실행하면
```python
c.x += 1 
Testclass2.y +=1 
print("현재 인스턴스에서 %s 회 출력" % c.x)
print("전체 인스턴스에서 총 %s 회 출력" % c.y)
```
이 실행되는데, 이때 Testclass2.y 이 정의되어 있지 않으므로 
```python
Testclass2.y +=1 
```
에서 에러가 난다. 

In [63]:
c.my_print()

AttributeError: type object 'Testclass2' has no attribute 'y'

### 예제 3

`-` 그렇다면 아래와 같이 수정하면 어떨까? 

In [65]:
class Testclass3: 
    def __init__(self):
        self.x=0
        Testclass3.y=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [67]:
a=Testclass3()
b=Testclass3()

In [68]:
a.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력


In [69]:
b.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 2 회 출력


In [70]:
a.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 3 회 출력


In [71]:
a.my_print()

현재 인스턴스에서 3 회 출력
전체 인스턴스에서 총 4 회 출력


In [72]:
b.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 5 회 출력


In [73]:
b.my_print()

현재 인스턴스에서 3 회 출력
전체 인스턴스에서 총 6 회 출력


`–` Testclass1과 동일한 기능이 수행되는것 같다. 

`-` 그런데 조금만 생각해보면 엉터리라는 것을 알 수 있다. 아래의 코드를 관찰하여보자. 

In [74]:
class Testclass3: 
    def __init__(self):
        self.x=0
        Testclass3.y=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
        
a=Testclass3()
a.my_print()
a.my_print()
b=Testclass3()
b.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 2 회 출력
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력


`-` Testclass3는 인스턴스를 생성할때마다 `y=0`이 설정된다. 그래서 
```python
b=Testclass3()
```
이 시점에서 의도하지 않게 '전체 인스턴스에서 출력된 횟수'를 의미하는 `y`가 초기화되었다. 

`-` 코드는 엉터리이지만, Testclass3은 의외로 분석할만한 가치가 있다. 특히 위의 실행결과를 시점별로 Testclass1과 비교해보면 재미있다. 

`-` Testclass1

```python
### Testclass1 
## 시점1: 클래스 오브젝트 생성 
class Testclass1: 
    x=0 
    y=0
    def my_print(self):
        self.x += 1 
        Testclass1.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
## 시점2: 인스턴스 오브젝트 a를 생성 
a=Testclass1()
## 시점3: a에서 메소드 실행 
a.my_print()
## 시점4: a에서 메소드를 한번 더 실행 
a.my_print()
## 시점5: 인스턴스 오브젝트 b를 생성
b=Testclass1()
## 시점6: b에서 메소드를 실행
b.my_print()
```

```
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 2 회 출력
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 3 회 출력
```

| |시점1 |시점2 |시점3 | 시점4 | 시점5| 시점6|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Testclass1.x|0| 0| 0| 0| 0| 0|
|Testclass1.y|0| 0| 1| 2| 2 | 3|
|a.x| 값없음     | 0| 1| 2| 2| 2|
|a.y| 값없음     | 0| 1| 2| 2| 3|
|b.x| 값없음|     값없음| 값없음| 값없음| 0| 1|
|b.y| 값없음|     값없음| 값없음| 값없음| 2| 3|

`–` Testclass3

```python
#### Testclass3
## 시점1: 클래스 오브젝트 생성 
class Testclass3: 
    def __init__(self):
        self.x=0
        Testclass3.y=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
## 시점2: 인스턴스 오브젝트 a를 생성 
a=Testclass3()
## 시점3: a에서 메소드 실행 
a.my_print()
## 시점4: a에서 메소드를 한번 더 실행 
a.my_print()
## 시점5: 인스턴스 오브젝트 b를 생성
b=Testclass3()
## 시점6: b에서 메소드를 실행
b.my_print()
```

```
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 2 회 출력
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
```

| |시점1 |시점2 |시점3 | 시점4 | 시점5| 시점6|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Testclass3.x|값없음| 값없음| 값없음| 값없음| 값없음| 값없음|
|Testclass3.y|값없음| 0| 1| 2| 0 | 1|
|a.x| 값없음     | 0| 1| 2| 2| 2|
|a.y| 값없음     | 0| 1| 2| 0| 1|
|b.x| 값없음|     값없음| 값없음| 값없음| 0| 1|
|b.y| 값없음|     값없음| 값없음| 값없음| 0| 1|

`–` Testclass3.y가 업데이트 되면 a.y, b.y도 자동으로 업데이트 된다. 

숙제: 

```python
#### Testclass3
## 시점1: 클래스 오브젝트 생성 
class Testclass3: 
        y=0
    def __init__(self):
        self.x=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
## 시점2: 인스턴스 오브젝트 a를 생성 
a=Testclass3()
## 시점3: a에서 메소드 실행 
a.my_print()
## 시점4: a에서 메소드를 한번 더 실행 
a.my_print()
## 시점5: 인스턴스 오브젝트 b를 생성
b=Testclass3()
## 시점6: b에서 메소드를 실행
b.my_print()
```

| |시점1 |시점2 |시점3 | 시점4 | 시점5| 시점6|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Testclass3.x| | | | | | |
|Testclass3.y|| | | | | |
|a.x| | | | | | |
|a.y| | | | | | |
|b.x| | | | | | |
|b.y| | | | | | |

## 네임스페이스

### 예제1

`-` 아래의 코드를 관찰하라. 

In [1]:
class Testclass1: 
    x=0

In [2]:
Testclass1.x

0

In [3]:
a=Testclass1()

In [4]:
a.x

0

`–` Testclass1.x를 수정하면 a.x가 강제로 수정된다. 

In [5]:
Testclass1.x=100

In [6]:
a.x

100

`-` a.x를 수정한다고 하여 Testclass1.x가 강제로 수정되는 것은 아님

In [7]:
a.x=200

In [8]:
Testclass1.x

100

In [9]:
a.x

200

`-` 이건 왜이러지?

In [10]:
Testclass1.x=300

In [11]:
a.x

200

`-` 아래의 상황과 비슷하다. 

In [32]:
## [code1]
x=39
def nextyear():
    y=x+1
    print(x,y)
nextyear()

39 40


In [29]:
## [code2]
x=39
def nextyear():
    y=x+1
    print(x,y)
    x=0
nextyear()

UnboundLocalError: local variable 'x' referenced before assignment

`–` [code1]은 잘 실행되던 코드다.

`-` [code2]는 실행안되는 코드다.

`-` [code2]와 [code1]의 차이점은 x=0이라는 코드가 추가로 포함되었는지 유무다. 

`–` (헛소리) x=0 이 잘못된 코드다!! 이걸 실행하는 과정에서 문제가 생겼다!! 

`-` (올바른소리) code1에서는 x는 global variable, code2에서는 x가 local variable 이라서 생기는 문제점이다. 

In [37]:
## [code2]
x=39
def nextyear():
    x=0
    y=x+1
    print(x,y)
nextyear()

0 1


In [38]:
x

39

`–` 다시 우리의 예제로 돌아오자.
```python
### 시점1
class Testclass1: 
    x=0
### 시점2
a=Testclass1()
### 시점3
Testclass1.x=100
### 시점4
a.x=200
### 시점5
Testclass1.x=300
```

| |시점1 |시점2 |시점3 | 시점4 | 시점5|
|:-:|:-:|:-:|:-:|:-:|:-:|
|Testclass1.x|0| 0| 100| 100| 300|
|a.x| 값없음     | 0| 100| 200| 200|
|a.x의 속성| -     | class| class| instance | instance|

`–` a.x가 클래스로부터 물려받은 속성인지 (그래서 클래스와 연결되어있는지) 아니면 instance가 독자적으로 가지고 있는 속성인지 어떻게 알 수 있을까? 

In [52]:
### 시점1
class Testclass1: 
    x=0
print('시점1',Testclass1.x)
### 시점2
a=Testclass1()
print('시점2',Testclass1.x,a.x,a.__dict__)
### 시점3
Testclass1.x=100
print('시점3',Testclass1.x,a.x,a.__dict__)
### 시점4
a.x=200
print('시점4',Testclass1.x,a.x,a.__dict__)
### 시점5
Testclass1.x=300
print('시점5',Testclass1.x,a.x,a.__dict__)

시점1 0
시점2 0 0 {}
시점3 100 100 {}
시점4 100 200 {'x': 200}
시점5 300 200 {'x': 200}


### 예제2

In [53]:
x=11 ## 전역변수 ... A

def f():
    x=22 ## 함수 f안에 설정된 지역변수 
    print(x) ## 전역에 x=11 있지만 함수안에 x=22가 있으므로 x=22를 사용. --> 22출력됨 

def g():
    print(x) ## 함수 g안에 x를 찾아봤는데 없음 --> 전역에서 x를 찾음 --> x=11 --> 11출력함. 

class Testclass2: 
    x=33 ## 클래스 변수 ... B
    def m1(self):
        x=44 ## 메소드 변수 ... C
    def m2(self):
        self.x=44 ## 인스턴스 변수 ... D 

`-` 결과를 관찰하고 해석해보자. 

In [54]:
print(x)

11


> Note: 전역변수 출력

In [55]:
f()

22


> Note: f에서 설정된 지역변수 22가 출력됨

In [56]:
x

11

> Note: f내의 지역변수를 사용하여도 전역변수는 변하지 않음. (함수내부에서 선언된 x=22는 함수외부에 영향을 주지못함)

In [57]:
g()

11


> Note: g에서 설정된 지역변수가 따로 없으므로 전역변수 출력

In [58]:
x,Testclass2.x

(11, 33)

> Note: 전역변수 x와 클래스오브젝트에 설정된 변수 x 

In [59]:
a=Testclass2()
(x,Testclass2.x,a.x),a.__dict__

((11, 33, 33), {})

> Note: 전역변수, 클래스 오브젝트내의 변수, 인스턴스내의 변수. `a.__dict__`의 결과로 보아 인스턴스내의 변수는 클래스 오브젝트내의 변수를 빌려쓰고 있다. 

In [60]:
Testclass2.x=200
(x,Testclass2.x,a.x),a.__dict__

((11, 200, 200), {})

> Note: 클래스오브젝트에서 변수를 고치면 인스턴스에 영향을 미침

In [61]:
a.m1()
(x,Testclass2.x,a.x),a.__dict__

((11, 200, 200), {})

> Note: 메소드 `m1`내에서 선언된 x=44라는 선언은 아무것도 변화시킬수 없음. 

In [62]:
a.m2()
(x,Testclass2.x,a.x),a.__dict__

((11, 200, 44), {'x': 44})

> Note: 메소드 `m2`에 있는 self.x는 결국 a.x라는 의미이고, 이 선언은 클래스오브젝트 내의 변수와 독립적으로 인스턴스오브젝트 내에서 통용되는 변수를 선언하는 것임. 이 선언의 결과는 `a.__dict__`의 출력결과에서도 확인가능.

In [63]:
Testclass2.x=300
(x,Testclass2.x,a.x),a.__dict__

((11, 300, 44), {'x': 44})

> Note: 이제는 a.x와 Testclass2.x 는 분리된 상태이므로, Testclass2.x의 값을 바꾸어도 a.x에는 값의 변화가 없음. 

`-` 뭐 어쩌라는 것임?

`-` 그냥 그렇다는것.. 내용을 모르고 사용한다면 예상하지 못한 오류가 발생할 수 있으므로 조심해서 사용하자. 

숙제: 전역변수(A), 클래스 변수(B), 메소드 변수(C), 인스턴스 변수(D)
    
답: A>B>D>C

## 연산자 오버로딩 

`-` 아래의 코드를 관찰하자. 

In [1]:
1+1

2

`-` 생각해보니까 `1`은 int class 에서 생성된 인스턴스이다. 

`-` 코드를 관찰하니 instance와 instance를 `+`라는 연산이 연결하는 형태임. 

In [169]:
class Student: 
    def __init__(self,age=20.0,semester=1):
        self.age=age
        self.semester=semester
    def __add__(self,val):
        # val==0: 휴학 
        # val==1: 등록
        if val==0: 
            self.age=self.age +0.5 
        elif val==1: 
            self.age=self.age+0.5
            self.semester=self.semester+1
        return self
    def __repr__(self):
        return '나이: %s \n학기: %s' % (self.age,self.semester)

In [170]:
guebin=Student()

In [76]:
guebin.age

20.0

In [77]:
guebin.semester

1

In [78]:
guebin

나이: 20.0 
학기: 1

In [79]:
guebin+1

나이: 20.5 
학기: 2

In [80]:
guebin+0

나이: 21.0 
학기: 2

In [81]:
guebin+0+0+0+0+1+0+1

나이: 24.5 
학기: 4

`-` 연산자 오버로드 핵심아이디어 
- 클래스가 일반 파이썬 연산을 재정의하는 것  
- 여기에서 연산은 단순히 더하기 빼기를 의미하는게 아니라, `print()`, `+`, `[0]` 와 같은 파이썬 내장문법을 모두 포괄하는 개념이라 이해하는 것이 옳다. 

In [192]:
guebin[0]

TypeError: 'Student' object is not subscriptable

In [183]:
class Student2(Student):
    def __getitem__(self,index):
        return [self.age,self.semester][index]

In [184]:
hynn=Student2()

In [185]:
hynn+1+1+0+0

나이: 22.0 
학기: 3

In [186]:
hynn[0]

22.0

In [187]:
hynn[1]

3

In [188]:
hynn[:]

[22.0, 3]

`-` 연산자 오버로딩을 이해하면 파이썬 전반에 대한 이해폭이 넓어진다. 

In [189]:
import pandas as pd 

In [190]:
df=pd.DataFrame({'age':[20,21.5],'semester':[1,2]})

In [191]:
df.iloc[:,0]

0    20.0
1    21.5
Name: age, dtype: float64

## 도움말 작성방법 

`-` 넘파이의 경우 아래와 같이 도움말이 잘 작성되어 있다. 

In [194]:
import numpy as np
a=np.array([1,2,3])
a?

Type:            ndarray
String form:     [1 2 3]
Length:          3
File:            ~/anaconda3/envs/py38r40/lib/python3.8/site-packages/numpy/__init__.py
Docstring:       <no docstring>
Class docstring:
ndarray(shape, dtype=float, buffer=None, offset=0,
        strides=None, order=None)

An array object represents a multidimensional, homogeneous array
of fixed-size items.  An associated data-type object describes the
format of each element in the array (its byte-order, how many bytes it
occupies in memory, whether it is an integer, a floating point number,
or something else, etc.)

Arrays should be constructed using `array`, `zeros` or `empty` (refer
to the See Also section below).  The parameters given here refer to
a low-level method (`ndarray(...)`) for instantiating an array.

For more information, refer to the `numpy` module and examine the
methods and attributes of an array.

Parameters
----------
(for the __new__ method; see Notes below)

shape : tuple of ints
    Shape of cr

`-` 하지만 우리는? 

In [196]:
hynn?

Type:        Student2
String form:
나이: 22.0 
학기: 3
Docstring:   <no docstring>


`-` 우리도 도움말을 작성하고 싶다. 

In [203]:
class Student2(Student):
    '''
    Student2는 Student의 개선 
    
    # Student 클래스의 기능 
    1. 출력기능 (__repr__)
    2. 연산기능 (__add__): 학기와 나이를 카운트 
    Examples
    --------
    >>> hynn=Student2()
    >>> hynn+1
    나이: 20.5 
    학기: 2
    
    # Student2에서 추가된 기능 
    1. 인덱싱
    '''
    def __getitem__(self,index):
        return [self.age,self.semester][index]

In [204]:
hynn=Student2()

In [205]:
hynn?

Type:        Student2
String form:
나이: 20.0 
학기: 1
Docstring:  
Student2는 Student의 개선 

# Student 클래스의 기능 
1. 출력기능 (__repr__)
2. 연산기능 (__add__): 학기와 나이를 카운트 
Examples
--------
>>> hynn=Student2()
>>> hynn+1
나이: 20.5 
학기: 2

# Student2에서 추가된 기능 
1. 인덱싱


In [206]:
hynn=Student2(21,1)

In [207]:
hynn

나이: 21 
학기: 1

In [208]:
hynn?

Type:        Student2
String form:
나이: 21 
학기: 1
Docstring:  
Student2는 Student의 개선 

# Student 클래스의 기능 
1. 출력기능 (__repr__)
2. 연산기능 (__add__): 학기와 나이를 카운트 
Examples
--------
>>> hynn=Student2()
>>> hynn+1
나이: 20.5 
학기: 2

# Student2에서 추가된 기능 
1. 인덱싱


## self에 대한 진실

`–` 사실 이름이 self가 아니어도 된다. 

In [212]:
class MooYaHo: 
    def __init__(a):
        a.text='mooyaho'

In [213]:
moo1=MooYaHo()

In [214]:
moo1.text

'mooyaho'

`–` 그런데 self를 많이 쓴다. 

숙제: MooYaHo class의 도움말에 학번을 넣고 출력 